serve/tests/unittest/test_engine_with_samplers.py

```diff
@@ -138,9 +138,53 @@ def test_ignore_eos(
         engine.stop()
 
 
+def test_stop(
+    model_artifact_path,
+    use_staging_engine,
+    max_num_batched_tokens=2560,
+    max_input_len=2560,
+    num_requests=5,
+):
+    prompt = "Write a merge sort program in Python."
+    engine = create_engine(
+        model_artifact_path,
+        use_staging_engine,
+        max_num_batched_tokens,
+        max_input_len,
+    )
+    ignore_eos = False
+    requests = []
+    for n, stop in enumerate(["\n", ["\n"], "\n\n", "!", ["n", "!"]]):
+        requests.append(create_request(idx=str(n), prompt=prompt, temp=0, max_tokens=300, stop=stop, ignore_eos=False))
+    engine.add(requests)
+
+    generated = ["" for _ in range(num_requests)]
+
+    while engine.has_pending_requests():
+        results = engine.step()
+        for res in results.outputs:
+            assert len(res.sequences) == 1
+            seq = res.sequences[0]
+            req_id = int(res.request_id)
+            if seq.is_finished:
+                # TODO: Currently staging engine returns FinishReason.Cancelled.
```

Comment:
@jroesch, do you think this would be a problem to fix immediately?

Reply:
No I think its fine if we return Cancelled if that's ok, would rather things are mostly functional